### PR TITLE
Leaves - Tiffany

### PR DESCRIPTION
**Pull request by TiffanyChio**
# Task List
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
Describe in your own words what the Model is doing in Rails | The model is like a Ruby class. It has states and methods.
Describe in your own words what the Controller is doing in Rails | The controller facilitates the actions or messages between browser, model, and view.
Describe in your own words what the View is doing in Rails | The view renders a screen on your browser. It is trying to present a page. It doesn’t manipulate your data.
Describe an edge-case controller test you wrote | “strong params will prevent updates to completion_date to be passed in” The use of strong parameters will not allow completion_date to be updated even if a params hash containing information on completion_date has been passed in to the patch method. 
What is the purpose of using strong params? (i.e. the params method in the controller) | Strong params is a way to dry up our code. We can’t feed the params hash directly to our create or update methods. This is a safeguard to prevent unscrupulous users from passing non-valid values. So strong params will allow us to write shorter code in a safe way.
How are Rails migrations related to Rails models? | Migrations describe how a model changes and can also add instances of a model.
Describe one area of Rails that are still unclear on | I don’t know. I definitely feel like there’s a lot that I don’t understand but I don’t know what I don’t understand yet.

### CODE REVIEW

File: app/controllers/tasks_controller.rb [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+    if @task.nil?
+      redirect_to root_path
+      return
+    end
+    
+    @task.destroy
+    redirect_to root_path
+  end
+  
+  def toggle_complete
```
Comment (beccaelenzil) [inline app/controllers/tasks_controller.rb:70]: Consider how having two separate actions: `toggle_complete` and `toggle_incomplete` would make this action idempotent.

File: test/controllers/tasks_controller_test.rb [hunk 2]
```diff
@@ -71,56 +65,154 @@
...
+      id = -1
+      
+      expect {
+        patch task_path(id), params: changes_hash
+      }.wont_change "Task.count"
+      
+      must_redirect_to root_path
+    end
+    
+    it "will not update if params hash do not include changes" do
```
Comment (beccaelenzil) [inline test/controllers/tasks_controller_test.rb:140]: Nice work including this test.

File: test/controllers/tasks_controller_test.rb [hunk 3]
```diff
@@ -71,56 +65,154 @@
...
+      expect(updated_task.completion_date).must_be_nil
     end
   end
-
+  
   # Complete these tests for Wave 4
   describe "destroy" do
-    # Your tests go here
-
+    it "can destroy a model" do
```
Comment (beccaelenzil) [inline test/controllers/tasks_controller_test.rb:172]: Make sure to test for an invalid id

**Comment by beccaelenzil** [comment]
# Task List
## What We're Looking For

Feature | Feedback
:------------- | :-------------
**Baseline** |
Appropriate Git Usage with no extraneous files checked in | check
Answered comprehension questions | check
Successfully handles: Index, Show | check
Index & Show tests pass |check
Successfully handles: New, Create | check
New & Create tests pass |check
Successfully handles: Edit, Update |check
Tests for Edit & Update test valid & invalid task ids | check
Successfully handles: Destroy, Task Complete |check
Tests for Destroy & Task Complete include tests for valid and invalid task ids | Make sure to test for an invalid id for Destroy 
Routes follow RESTful conventions | check
Uses named routes (like `_path`) | check
**Overall** | Good job on Task List! It functions as expected. You hit all of the learning goals: using partials, using strong params, and following Rails conventions.
